### PR TITLE
[WIP] InProcessEmitToolchain

### PR DESCRIPTION
**Pull request by ig-sinicyn**
See #919 for initial proposal details.

How it works:

1. Main logic starts from the `RunnableEmitter.EmitPartitionAssembly()`. The Runnable emitter emits MSIL code that closely matches code produced by `RoslynToolchain`, assembly name is %ProgramName%Emitted.dll
1. Bootstrap part for triggering the build and executing the benchmarks is located in InProcessEmitToolchain classes (it may be renamed to InProcessToolchain if you decide to drop the old one)
1. There is prooftest `InProcessBenchmarkEmitsSameMsil` that compares code generated by `RoslynToolchain` and `InProcessToolchain` and allows to keep these two in sync. NB: .Net core does not support emitted assemblies saving, so the test works only for full .net; on other platforms the test runs benchmarks but does not perform the diff.
1. The diff is performed by NaiveRunnableEmitDiff helper class. The helper uses fail-fast approach and does not provide helpful diagnostics. I've tried to improve the output but it did not work for me. It turns out it is much easier to use IlSpy for troubleshooting so I decided to not invest in prettifying the diff check.
1. Additionally, I've added checks for InProcessEmitToolchain into `ArgumentsTests` and `MemoryDiagnoserTests`. Had to fix the MemoryDiagnoserTests, there was memleak due to the xUnit output.

Ok, the code.
Benchmark:
```c#
        private int refValueHolder;

        [Benchmark, Arguments(123.0, 4, "5", null)]
        public ref int RefReturnManyArgsCase(ref double i, int j, string k, object l)
        {
            Thread.Sleep(100);
            return ref refValueHolder;
        }
```

RoslynToolchain:
```IL
.method public hidebysig 
	instance int32& __ForDisassemblyDiagnoser__ () cil managed noinlining 
{
	// Method begins at RVA 0x6ee8
	// Code size 56 (0x38)
	.maxstack 5
	.locals init (
		[0] float64&,
		[1] int32,
		[2] string,
		[3] object
	)

	// if (NotEleven == 11)
	IL_0000: ldarg.0
	IL_0001: ldfld int32 BenchmarkDotNet.Autogenerated.Runnable_4::NotEleven
	IL_0006: ldc.i4.s 11
	IL_0008: bne.un.s IL_0031

	// ref double _argField = ref __argField0;
	IL_000a: ldarg.0
	IL_000b: ldflda float64 BenchmarkDotNet.Autogenerated.Runnable_4::__argField0
	IL_0010: stloc.0
	// int _argField2 = __argField1;
	IL_0011: ldarg.0
	IL_0012: ldfld int32 BenchmarkDotNet.Autogenerated.Runnable_4::__argField1
	IL_0017: stloc.1
	// string _argField3 = __argField2;
	IL_0018: ldarg.0
	IL_0019: ldfld string BenchmarkDotNet.Autogenerated.Runnable_4::__argField2
	IL_001e: stloc.2
	// object _argField4 = __argField3;
	IL_001f: ldarg.0
	IL_0020: ldfld object BenchmarkDotNet.Autogenerated.Runnable_4::__argField3
	IL_0025: stloc.3
	// return ref RefReturnManyArgsCase(ref _argField, _argField2, _argField3, _argField4);
	IL_0026: ldarg.0
	IL_0027: ldloc.0
	IL_0028: ldloc.1
	IL_0029: ldloc.2
	IL_002a: ldloc.3
	IL_002b: call instance int32& [BenchmarkDotNet.IntegrationTests]BenchmarkDotNet.IntegrationTests.InProcess.EmitTests.SampleBenchmark::RefReturnManyArgsCase(float64&, int32, string, object)
	// (no C# code)
	IL_0030: ret

	// return ref workloadDefaultValueHolder;
	IL_0031: ldarg.0
	IL_0032: ldflda int32 BenchmarkDotNet.Autogenerated.Runnable_4::workloadDefaultValueHolder
	// (no C# code)
	IL_0037: ret
} // end of method Runnable_4::__ForDisassemblyDiagnoser__
```

InProcessEmitToolchain:
```IL
.method public hidebysig 
	instance int32& __ForDisassemblyDiagnoser__ () cil managed noinlining 
{
	// Method begins at RVA 0x656c
	// Code size 59 (0x3b)
	.maxstack 6
	.locals init (
		[0] float64&,
		[1] int32,
		[2] string,
		[3] object
	)

	// if (NotEleven == 11)
	IL_0000: ldarg.0
	IL_0001: ldfld int32 BenchmarkDotNet.Autogenerated.Runnable_4::NotEleven
	IL_0006: ldc.i4.s 11
	IL_0008: bne.un IL_0034

	// ref double _argField = ref __argField0;
	IL_000d: ldarg.0
	IL_000e: ldflda float64 BenchmarkDotNet.Autogenerated.Runnable_4::__argField0
	IL_0013: stloc.0
	// int _argField2 = __argField1;
	IL_0014: ldarg.0
	IL_0015: ldfld int32 BenchmarkDotNet.Autogenerated.Runnable_4::__argField1
	IL_001a: stloc.1
	// string _argField3 = __argField2;
	IL_001b: ldarg.0
	IL_001c: ldfld string BenchmarkDotNet.Autogenerated.Runnable_4::__argField2
	IL_0021: stloc.2
	// object _argField4 = __argField3;
	IL_0022: ldarg.0
	IL_0023: ldfld object BenchmarkDotNet.Autogenerated.Runnable_4::__argField3
	IL_0028: stloc.3
	// return ref RefReturnManyArgsCase(ref _argField, _argField2, _argField3, _argField4);
	IL_0029: ldarg.0
	IL_002a: ldloc.0
	IL_002b: ldloc.1
	IL_002c: ldloc.2
	IL_002d: ldloc.3
	IL_002e: call instance int32& [BenchmarkDotNet.IntegrationTests]BenchmarkDotNet.IntegrationTests.InProcess.EmitTests.SampleBenchmark::RefReturnManyArgsCase(float64&, int32, string, object)
	// (no C# code)
	IL_0033: ret

	// return ref workloadDefaultValueHolder;
	IL_0034: ldarg.0
	IL_0035: ldflda int32 BenchmarkDotNet.Autogenerated.Runnable_4::workloadDefaultValueHolder
	// (no C# code)
	IL_003a: ret
} // end of method Runnable_4::__ForDisassemblyDiagnoser__
```

### CODE REVIEW

**Comment by ig-sinicyn** [comment]
@adamsitnik, @AndreyAkinshin  There's something wrong with RoslynToolchain tests on .Net Core.

May you look at `CheckIfRoslynToolchainFailsOnNetCore` test?

File: src/BenchmarkDotNet/Helpers/Reflection.Emit/IlGeneratorEmitOpExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+                    ilBuilder.Emit(OpCodes.Ldarg_2);
+                    break;
+                case 3:
+                    ilBuilder.Emit(OpCodes.Ldarg_3);
+                    break;
+                case var i when i < 255:
+                    ilBuilder.Emit(OpCodes.Ldarg_S, (byte)position);
+                    break;
+                default:
+                    ilBuilder.Emit(OpCodes.Ldarg, position);
```
[the remark this answers is not in the document]
Comment (Warpten) [reply src/BenchmarkDotNet/Helpers/Reflection.Emit/IlGeneratorEmitOpExtensions.cs]: (You should wire in additional checks on the amount of local variables - MSIL only allows up to 65536.)

I (imvho) think it'd be nicer to write this as expression trees as they take away the work of emitting IL from your hands, albeit with some extra overhead. Not everyone is well versed in IL which itself is full of tricks (you faced issues with nops being inserted - imagine what happens when you pass in arguments that take less bytes than what the opcode expects!)

However, I'm not sure about maintainability ; this is quite a large chunk of code, no matter how it's implemented. The only actual benefit from expressions is that they failover relatively fast, meaning you would see your errors faster. For IL, I can recommend [Sigil](https://github.com/kevin-montrose/Sigil) (although it seems unmaintained currently), which does a pretty good job at that (at the cost of not running on .NET Core, and being quite a bit slower for large methods.

File: src/BenchmarkDotNet/Helpers/Reflection.Emit/IlGeneratorEmitOpExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+                    ilBuilder.Emit(OpCodes.Stloc_2);
+                    break;
+                case 3:
+                    ilBuilder.Emit(OpCodes.Stloc_3);
+                    break;
+                case var i when i < 255:
+                    ilBuilder.Emit(OpCodes.Stloc_S, (byte)local.LocalIndex);
+                    break;
+                default:
+                    ilBuilder.Emit(OpCodes.Stloc, local.LocalIndex);
```
[the remark this answers is not in the document]
Comment (ig-sinicyn) [reply src/BenchmarkDotNet/Helpers/Reflection.Emit/IlGeneratorEmitOpExtensions.cs]: If I do the output diffs from RoslynToolchain.

[Paste this](https://gist.github.com/ig-sinicyn/8d8e7f185b326949d1c7e814aeb95470) into https://sharplab.io/ (they remove share buttons). Output is
```
        IL_07ed: ldarg.0
        IL_07ee: ldfld uint8 BenchmarkDotNet.Autogenerated.Runnable_0::__argField254
        IL_07f3: stloc.s 254
        IL_07f5: ldarg.0
        IL_07f6: ldfld uint8 BenchmarkDotNet.Autogenerated.Runnable_0::__argField255
        IL_07fb: stloc 255
        IL_07ff: nop
        IL_0800: nop
        IL_0801: ldarg.0
        IL_0802: ldfld uint8 BenchmarkDotNet.Autogenerated.Runnable_0::__argField256
        IL_0807: stloc 256
        IL_080b: nop
        IL_080c: nop
        IL_080d: ldarg.0
        IL_080e: ldfld uint8 BenchmarkDotNet.Autogenerated.Runnable_0::__argField257
        IL_0813: stloc 257
        IL_0817: nop
        IL_0818: nop
```
Surprisingly, there are no NOPs for args with index > 512

Other helper method use int overload for same reason

[the remark this answers is not in the document]
Comment (ig-sinicyn) [reply src/BenchmarkDotNet/Helpers/Reflection.Emit/IlGeneratorEmitOpExtensions.cs]: I've think twice and decide it's better to emit correct IL than try to mimic Roslyn output. 
Will add the casts.

**Comment by ig-sinicyn** [comment]
> I (imvho) think it'd be nicer to write this as expression trees as they take away the work of emitting IL from your hands, albeit with some extra overhead. Not everyone is well versed in IL which itself is full of tricks (you faced issues with nops being inserted - imagine what happens when you pass in arguments that take up less bytes than what the opcode expects!)

First attempt was made this way. Sadly it did not provide same IL output the RoslynToolchain do and the result was unverifiable.

> However, I'm not sure about maintainability ; this is quite a large chunk of code, no matter how it's implemented.

Maintanability is not so hard due to the diff tests (see `InProcessBenchmarkEmitsSameMsil` cases).
If the output of both toolchains matches there is no need to dig into toolchain code.
If there are differences it is easy to fix them as the code is pretty simple, nothing more than a long sequence of `.Emit(OpCode)` calls annotated with IL copied from Roslyn toolchain output.

**Comment by adamsitnik** [comment]
@ig-sinicyn you have done a tremendous job here! the PR is really BIG, I am going to review it over the weekend. It's going to be our main feature of 0.11.3!

PS. I am also glad you stopped using `#regions` :)

**Comment by ig-sinicyn** [comment]
Okay, attempt  2:)

@adamsitnik , @AndreyAkinshin, can you look at this?
I need help with the CheckIfRoslynToolchainFailsOnNetCore, it looks like there's something wrong with my config for it.

**Comment by ig-sinicyn** [comment]
Ping?:)

**Comment by AndreyAkinshin** [comment]
@ig-sinicyn sorry for the lack of response. Adam and I where pretty busy lately and your pull request is a pretty big, so it can't be quickly reviewed. We really appreciate your efforts and we are going to review it as soon as we find some time for that.

**Comment by ig-sinicyn** [comment]
I've rebased the branch onto actual master, no conflicts so far.

**Comment by adamsitnik** [comment]
@ig-sinicyn  the code is perfect, I am going to merge the PR to a new branch, fix the build and test issues, test it and then merge it to master

**Comment by ig-sinicyn** [comment]
@adamsitnik 
I've merged master into ig-sinicyn:feature-inprocessemit and going to fix broken tests.

Does BDN vnext support static benchmark methods? If yes, what is proper way to test them? 
Currently tests fail due to the CompilationValidator.ValidateBindingModifiers() check.

If no, should I remove support for static members from the emitter?

**Pull request merged**